src/main.rs

```diff
@@ -103,7 +103,7 @@ async fn main(spawner: Spawner) {
     let blinky_fut = blinky.run();
     // Run everything concurrently.
     // If we had made everything `'static` above instead, we could do this using separate tasks instead.
-    join(join(usb_fut, usb_mass_storage_fut), blinky_fut).await;
+    join_array([usb_fut, usb_mass_storage_fut, blinky_fut]).await;
```

Comment:
Needs reverting - Future's of inconsistent types :(

Reply:
We can make a few little changes to sort this - I can suggest here or open a PR to your repo/this branch, any preference?

Reply:
I've reverted for now. Happy for you to suggest here.

Reply:
Sweet, for the `()` vs `!`, we can propagate the never out from bbb so they're all `Future<Output = !>`:

```diff
diff --git a/src/scsi/mod.rs b/src/scsi/mod.rs
@@ -85,9 +85,9 @@ impl<'d, D: Driver<'d>, M: RawMutex> Scsi<'d, D, M> {
         }
     }
 
-    pub async fn run(&mut self, handler: &mut impl Handler) {
+    pub async fn run(&mut self, handler: &mut impl Handler) -> ! {
         let mut adapter = Adapter { handler };
-        self.transport.run(&mut adapter).await;
+        self.transport.run(&mut adapter).await
     }
 }
 
diff --git a/src/usb_mass_storage/mod.rs b/src/usb_mass_storage/mod.rs
@@ -81,7 +81,7 @@ impl<'d, D: Driver<'d>, M: RawMutex> UsbMassStorage<'d, D, M> {
         }
     }
 
-    pub async fn run(&mut self, handler: &mut impl Handler) {
+    pub async fn run(&mut self, handler: &mut impl Handler) -> ! {
         self.scsi.run(handler).await
     }
 }
```


---

Also If you like, for this bit we can use `join3`:

```rust
join3(usb_fut, usb_mass_storage_fut, blinky_fut).await;
```